### PR TITLE
fix(server): constrain OpenAPI route module keys

### LINKED ISSUE #3357: Constrain OpenAPI route keys before dynamic route-module imports
# Constrain OpenAPI route keys before dynamic route-module imports

## Summary

`src/server.ts` auto-registers routes from generated OpenAPI artifacts under `artifacts/codex/`. The route-module key is derived from the OpenAPI path plus the OpenAPI method name, but only the path portion is normalized. The method name is appended before the code checks whether it is an expected HTTP verb, and the resulting key is used in filesystem checks and dynamic imports.

This means generated OpenAPI content can influence module resolution outside the intended `src/routes` boundary or remap an existing handler to an unintended route path through lossy key collisions.

## Security impact

A compromised generation step or malicious repository contribution that controls the generated OpenAPI artifact could cause server startup to probe/import an unintended local module or bind an existing route handler to an unexpected path. The practical blast radius depends on which generated artifact is deployed and which modules exist outside `src/routes`.

This draft intentionally avoids exact traversal strings and step-by-step reproduction details because the repository is public.

## Evidence

- `repo/src/server.ts:82-93` reads and parses `artifacts/codex/openapi.json` or `artifacts/codex/openapi.yaml`.
- `repo/src/server.ts:96-100` iterates OpenAPI paths/method names and derives the route-module key.
- `repo/src/server.ts:63-72` uses the derived key in `path.join()` and dynamic `import()` calls under `./routes/`.
- `repo/src/server.ts:103-133` checks the HTTP method only after attempting to load the module.
- `repo/src/server.ts:102` builds the registered Fastify path from the original OpenAPI path, not from an allowlisted route manifest.

## Recommended remediation

- Validate the method name against an exact allowlist before constructing or loading a route-module key.
- Reject any final key containing path separators, dot segments, empty segments, or characters outside a strict route-module identifier pattern.
- Resolve the candidate module path and verify it remains under the intended `src/routes` directory before importing it.
- Prefer a generated manifest or explicit `operationId` mapping over lossy path-derived module names.
- Validate generated OpenAPI paths against expected public routes and make duplicate/invalid route registration fail with a controlled diagnostic.
- Add regression tests for unexpected method names, path separators, dot segments, key collisions, and duplicate routes.

### PR DESCRIPTION
**Pull request by ootakazuhiko**
## Summary
- Replace path-derived OpenAPI route-module keys with an explicit allowlisted route manifest.
- Validate OpenAPI methods and path segments before any route key or dynamic import is resolved.
- Resolve route modules through contained filesystem paths under `src/routes` and import via file URLs.
- Add unit coverage for unsafe methods, unsafe paths, unexpected routes, and route-module path containment.

Closes #3357

## Acceptance
- OpenAPI auto-registration only accepts the known public route/method combinations for reservations and inventory.
- Unsupported methods, unsafe paths, and unexpected route definitions fail closed before module-key construction.
- Dynamic route imports are resolved under `src/routes` and reject unsafe module keys or path escapes.
- Regression tests cover the route manifest and path-containment behavior.

## Verification
- `pnpm -s exec vitest run tests/unit/server/openapi-route-security.test.ts --reporter dot`
- `pnpm -s exec eslint --quiet src/server.ts src/server/openapi-route-security.ts tests/unit/server/openapi-route-security.test.ts`
- `pnpm -s run types:check`
- `pnpm -s run build`
- `pnpm -s run check:schemas`
- `pnpm -s run check:doc-consistency`
- `git diff --check`

## Rollback
- Revert this PR to restore the previous OpenAPI auto-registration behavior. If reverted, avoid using untrusted OpenAPI artifacts for route auto-registration until an equivalent allowlist/path-containment control is in place.

### CODE REVIEW

File: src/server/openapi-route-security.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+export function collectOpenApiRouteDescriptors(spec: OpenApiLikeSpec): OpenApiRouteDescriptor[] {
+  const descriptors: OpenApiRouteDescriptor[] = []
+  const seenRoutes = new Set<string>()
+  const seenModuleKeys = new Set<string>()
+
+  for (const [openApiPath, methods] of Object.entries(spec.paths ?? {})) {
+    const methodMap = toPlainObject(methods)
+    for (const [rawMethod] of Object.entries(methodMap)) {
+      const descriptor = resolveOpenApiRouteDescriptor(openApiPath, rawMethod)
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply src/server/openapi-route-security.ts:93]: 対応しました。OpenAPI Path Item のメタデータフィールド（`$ref` / `summary` / `description` / `servers` / `parameters`）は descriptor 解決前にスキップするようにし、共有 `parameters` 等を含む正規の Path Item が起動時失敗にならないよう修正しました。併せて、`head` のような未サポート operation key は引き続き fail-closed になる回帰テストを追加しています。

File: src/server.ts [hunk 2]
```diff
@@ -1,27 +1,28 @@
...
+  resolveRouteModulePath,
+  type OpenApiLikeSpec,
+} from './server/openapi-route-security.js'
 import { handler as postReservation } from './routes/reservations-post.js'
 import { handler as deleteReservation } from './routes/reservations-id-delete.js'
 import { handler as getInventory } from './routes/inventory-sku-get.js'
 
 const server = Fastify({ logger: true })
+const moduleDir = path.dirname(fileURLToPath(import.meta.url))
+const routesDir = path.resolve(moduleDir, 'routes')
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply src/server.ts:18]: 対応しました。`require.main === module` を廃止し、`process.argv[1]` と `import.meta.url` / `pathToFileURL()` を比較する ESM-safe な entrypoint guard に変更しました。`pnpm -s run build` 後に `node -e "import('./dist/src/server.js')..."` で module evaluation が成功することも確認しています。
